test/EngineTest.cpp

```diff
@@ -23,114 +32,259 @@ auto I = [](const auto& id) {
   }
-  a.push_back({I(4000001), I(200000)});
-  b.push_back({I(4000001), I(200000)});
-  CALL_FIXED_SIZE((std::array{lwidth, rwidth, reswidth}), &Join::join, &J, a, 0,
-                  b, 0, &res);
-  ASSERT_EQ(2u, res.size());
-
-  a.clear();
-  b.clear();
-  res.clear();
-
-  IdTable c(1, allocator());
-  c.push_back({I(0)});
-
-  b.push_back({I(0), I(1)});
-  b.push_back({I(0), I(2)});
-  b.push_back({I(1), I(3)});
-  b.push_back({I(1), I(4)});
-
-  lwidth = b.numColumns();
-  rwidth = c.numColumns();
-  reswidth = b.numColumns() + c.numColumns() - 1;
-  // reset the IdTable.
-  res = IdTable(reswidth, allocator());
-  CALL_FIXED_SIZE((std::array{lwidth, rwidth, reswidth}), &Join::join, J, b, 0,
-                  c, 0, &res);
-
-  ASSERT_EQ(2u, res.size());
-
-  ASSERT_EQ(I(0), res(0, 0));
-  ASSERT_EQ(I(1), res(0, 1));
-
-  ASSERT_EQ(I(0), res(1, 0));
-  ASSERT_EQ(I(2), res(1, 1));
+  leftIdTable.push_back({400000, 200000});
+  rightIdTable.push_back({400000, 200000});
+  myTestSet.push_back(normalJoinTest(makeIdTableFromVector(leftIdTable), 0,
+        makeIdTableFromVector(rightIdTable), 0, makeIdTableFromVector(sampleSolution), true));
+  
+  leftIdTable = {};
+  rightIdTable = {};
+  sampleSolution = {{40000, 200000, 200000}, {4000001, 200000, 200000}};
+  for (size_t i = 1; i <= 10000; ++i) {
+    leftIdTable.push_back({4 + i, 2 + i});
+  }
+  leftIdTable.push_back({40000, 200000});
+  rightIdTable.push_back({40000, 200000});
+  for (size_t i = 1; i <= 10000; ++i) {
+    leftIdTable.push_back({40000 + i, 2 + i});
+  }
+  leftIdTable.push_back({4000001, 200000});
+  rightIdTable.push_back({4000001, 200000});
+  myTestSet.push_back(normalJoinTest(makeIdTableFromVector(leftIdTable), 0,
+        makeIdTableFromVector(rightIdTable), 0, makeIdTableFromVector(sampleSolution), true));
+  
+  leftIdTable = {{0, 1}, {0, 2}, {1, 3}, {1, 4}};
+  rightIdTable = {{0}};
+  sampleSolution = {{0, 1}, {0, 2}};
+  myTestSet.push_back(normalJoinTest(makeIdTableFromVector(leftIdTable), 0,
+        makeIdTableFromVector(rightIdTable), 0, makeIdTableFromVector(sampleSolution), true));
+
+  return myTestSet;
+}
+
+
+TEST(EngineTest, joinTest) {
+  Join J{Join::InvalidOnlyForTestingJoinTag{}};
+  auto JoinLambda = [&J]<int A, int B, int C>(auto&&... args) {
+    return J.join<A, B, C>(AD_FWD(args)...);
+  };
+  goThroughSetOfTestsWithJoinFunction(createNormalJoinTestSet(), JoinLambda);
 };
 
 TEST(EngineTest, optionalJoinTest) {
-  IdTable a(3, allocator());
-  a.push_back({I(4), I(1), I(2)});
-  a.push_back({I(2), I(1), I(3)});
-  a.push_back({I(1), I(1), I(4)});
-  a.push_back({I(2), I(2), I(1)});
-  a.push_back({I(1), I(3), I(1)});
-  IdTable b(3, allocator());
-  b.push_back({I(3), I(3), I(1)});
-  b.push_back({I(1), I(8), I(1)});
-  b.push_back({I(4), I(2), I(2)});
-  b.push_back({I(1), I(1), I(3)});
+  IdTable a = makeIdTableFromVector(std::vector<std::vector<size_t>>{{{4, 1, 2}, {2, 1, 3}, {1, 1, 4}, {2, 2, 1}, {1, 3, 1}}});
```

Comment:
```suggestion
  IdTable a = makeIdTableFromVector({{4, 1, 2}, {2, 1, 3}, {1, 1, 4}, {2, 2, 1}, {1, 3, 1}});
```
Should work here (and in all the other places below), because your function `makeIdTableFromVector` is  not
overloaded.